File: sdk/src/main/java/org/coralibre/android/sdk/internal/device_info/DeviceInfo.java
package org.coralibre.android.sdk.internal.device_info;

import static org.coralibre.android.sdk.internal.device_info.ConfidenceLevel.getConfidenceLevel;

public class DeviceInfo implements Comparable<DeviceInfo> {

    private String manufacturer;
    private String device;
    private String model;
    private int rssiCorrection;
    private int tx;
    private ConfidenceLevel calibrationConfidence;

    public DeviceInfo(String manufacturer,
                      String device,
                      String model,
                      int rssiCorrection,
                      int tx,
                      ConfidenceLevel calibrationConfidence) {
        this.manufacturer = manufacturer;
        this.device = device;
        this.model = model;
        this.rssiCorrection = rssiCorrection;
        this.tx = tx;
        this.calibrationConfidence = calibrationConfidence;
    }

    public DeviceInfo(String rawString) {
        String[] row = rawString.split(",");
        this.manufacturer = row[0];
        this.device = row[1];
        this.model = row[2];
        this.rssiCorrection = Integer.parseInt(row[3]);
        this.tx = Integer.parseInt(row[4]);
        try {
            this.calibrationConfidence = getConfidenceLevel(Integer.parseInt(row[5]));
        } catch (NumberFormatException e) {
            this.calibrationConfidence = getConfidenceLevel(row[5]);
        }
    }

    public String getManufacturer() {
        return manufacturer;
    }

    public String getDevice() {
        return device;
    }

    public String getModel() {
        return model;
    }

    public int getRssiCorrection() {
        return rssiCorrection;
    }

    public int getTx() {
        return tx;
    }

    public ConfidenceLevel getCalibrationConfidence() {
        return calibrationConfidence;
    }

    @Override
    public String toString() {
        return String.format("%s,%s,%s,%d,%d,%s",
                manufacturer, device, model, rssiCorrection, tx, calibrationConfidence);
    }

    @Override
    public boolean equals(Object otherInfo) {
        if (otherInfo.getClass() != getClass()) return false;
        DeviceInfo otherDeviceInfo = (DeviceInfo) otherInfo;
        return otherDeviceInfo.getManufacturer().equalsIgnoreCase(manufacturer)
                && otherDeviceInfo.getDevice().equalsIgnoreCase(device)
                && otherDeviceInfo.getModel().equalsIgnoreCase(model)
                && otherDeviceInfo.getRssiCorrection() == rssiCorrection
                && otherDeviceInfo.getTx() == tx
                && otherDeviceInfo.getCalibrationConfidence() == calibrationConfidence;
    }

    @Override
    public int compareTo(DeviceInfo otherInfo) {
        if (manufacturer.compareToIgnoreCase(otherInfo.manufacturer) != 0) {
            return manufacturer.compareToIgnoreCase(otherInfo.manufacturer);
        }
        if (model.compareToIgnoreCase(otherInfo.model) != 0) {
            return model.compareToIgnoreCase(otherInfo.model);
        }
        if (device.compareToIgnoreCase(otherInfo.device) != 0) {
            return device.compareToIgnoreCase(otherInfo.device);
        }
        return 0;
    }
}

File: sdk/src/main/java/org/coralibre/android/sdk/internal/deviceinfo/DeviceInfo.kt
package org.coralibre.android.sdk.internal.deviceinfo

import org.coralibre.android.sdk.internal.deviceinfo.ConfidenceLevel.Companion.getConfidenceLevel
import java.util.Locale

class DeviceInfo : Comparable<DeviceInfo> {
    val manufacturer: String
    val device: String
    val model: String
    val rssiCorrection: Int
    val tx: Int
    var calibrationConfidence: ConfidenceLevel
        private set

    constructor(
        manufacturer: String,
        device: String,
        model: String,
        rssiCorrection: Int,
        tx: Int,
        calibrationConfidence: ConfidenceLevel
    ) {
        this.manufacturer = manufacturer
        this.device = device
        this.model = model
        this.rssiCorrection = rssiCorrection
        this.tx = tx
        this.calibrationConfidence = calibrationConfidence
    }

    constructor(rawString: String) {
        val row = rawString.split(",".toRegex()).toTypedArray()
        manufacturer = row[0]
        device = row[1]
        model = row[2]
        rssiCorrection = row[3].toInt()
        tx = row[4].toInt()
        try {
            calibrationConfidence = getConfidenceLevel(row[5].toInt())
        } catch (e: NumberFormatException) {
            calibrationConfidence = getConfidenceLevel(row[5])
        }
    }

    override fun toString(): String {
        return String.format(
            "%s,%s,%s,%d,%d,%s",
            manufacturer, device, model, rssiCorrection, tx, calibrationConfidence
        )
    }

    override fun equals(other: Any?): Boolean {
        if (other?.javaClass != javaClass) return false
        val otherDeviceInfo = other as DeviceInfo
        return (
            otherDeviceInfo.manufacturer.equals(manufacturer, ignoreCase = true) &&
                otherDeviceInfo.device.equals(device, ignoreCase = true) &&
                otherDeviceInfo.model.equals(model, ignoreCase = true) &&
                otherDeviceInfo.rssiCorrection == rssiCorrection &&
                otherDeviceInfo.tx == tx &&
                otherDeviceInfo.calibrationConfidence === calibrationConfidence
            )
    }

    override fun hashCode(): Int {
        val locale = Locale.ROOT
        var result = manufacturer.toLowerCase(locale).hashCode()
        result = 31 * result + device.toLowerCase(locale).hashCode()
        result = 31 * result + model.toLowerCase(locale).hashCode()
        result = 31 * result + rssiCorrection
        result = 31 * result + tx
        result = 31 * result + calibrationConfidence.hashCode()
        return result
    }

    override fun compareTo(other: DeviceInfo): Int {
        return compareBy<DeviceInfo>(
            { it.manufacturer.toLowerCase(Locale.ROOT) },
            { it.model.toLowerCase(Locale.ROOT) },
            { it.device.toLowerCase(Locale.ROOT) }
        ).compare(this, other)
    }
}
